Full name section Vuetified

Pytal opened the pull request:
### To Do

- [x]  Pass initial state from server to client
- [x]  Create service
- [x]  Abstract shared components
- [x]  Rewrite Email section
- [x]  Rewrite php markup and functionality in Vue `apps/settings/templates/settings/personal/persona.info.php`
- [x]  Rewrite with functional paradigm
- [x]  Refine input validation

### Note: The `TODO` comments are out-of-scope for this PR and will be addressed separately in future PRs

Contributes to #27869

Pytal committed:
Provide initial state

Signed-off-by: Christopher Ng <chrng8@gmail.com>

Pytal committed:
Create display name service and update constants

Signed-off-by: Christopher Ng <chrng8@gmail.com>

skjnldsv commented on apps/settings/templates/settings/personal/personal.info.php:
```diff
@@ -99,32 +99,7 @@
-				<span class="icon-error hidden" ></span>
-				<input type="hidden" id="displaynamescope" value="<?php p($_['displayNameScope']) ?>">
-			</form>
+			<div id="vue-displaynamesection" class="section"></div>
```
I think it would be cleaner to entirely move the whole page to vue now.
Meaning no more single-section injection on this template, just one big `ProfilePage` app like planned

Pytal replied:
This PR involved quite a bit of rewriting to abstract the shared components, so to keep this PR small I'll keep this one to just this one section — with the entire page in Vue coming up :)

artonge replied:
So, should we review, or wait for the other PR ?

Pytal replied:
Yeah it's good to review :)

Pytal commented:
/rebase

artonge commented on apps/settings/src/components/PersonalInfo/DisplayNameSection/DisplayName.vue:
```diff
@@ -0,0 +1,179 @@
+	methods: {
+		onDisplayNameChange(e) {
+			this.$emit('update:displayName', e.target.value.trim())
+			// $nextTick() ensures that references to this.dipslayName further down the chain give the correct non-outdated value
+			this.$nextTick(() => this.debounceDisplayNameChange())
```
Still dislike the `$nextTick` logic :(

Pytal replied:
Unforetunately without it the values may be sporadically out-of-date, am open to other solutions though :)

Pytal replied:
Unfortunately without it the values may be sporadically out-of-date, am open to other solutions though :)

artonge replied:
I would favor using the values from the event to make the function pure.

```js
onDisplayNameChange(e) {
			this.$emit('update:displayName', e.target.value.trim())
			this.debounceDisplayNameChange(e.target.value.trim())
},
```

Pytal replied:
I would be fine with moving to a functional paradigm as well, although from what I've seen from the rest of our codebase and as recommended by @skjnldsv we're using a lot of `this.[propname]` in our Vue methods, so I think it would be best at least for now to keep it consistent :)

Pytal replied:
I would be fine with moving to a functional paradigm as well, although from what I've seen from the rest of our codebase and as recommended by @skjnldsv we're using a lot of `this.[propname]` in our Vue methods, so I think it would be best at least for now to keep it consistent :) is that fine with you?

artonge replied:
Not that much, but if it is consistent with the rest of the codebase I guess I can look away when clicking Approve :)

Pytal replied:
@skjnldsv should we keep it consistent or start writing in a more functional paradigm?

Pytal replied:
@skjnldsv should we keep it consistent or start moving towards a more functional paradigm?

ChristophWurst replied:
:+1: on more functional programming where it's practical

skjnldsv replied:
> as recommended by @skjnldsv we're using a lot of `this.[propname]` in our Vue methods

Did I say that? I've been recommending the sync modifier approach quite a lot :wink: 
`nextTick` should indeed be a last resort :)

artonge commented on apps/settings/src/components/PersonalInfo/EmailSection/EmailSection.vue:
```diff
@@ -54,14 +58,14 @@
+import { ACCOUNT_PROPERTY_READABLE_ENUM, DEFAULT_ADDITIONAL_EMAIL_SCOPE } from '../../../constants/AccountPropertyConstants'
+import { savePrimaryEmail, savePrimaryEmailScope, removeAdditionalEmail } from '../../../service/PersonalInfo/EmailService'
+
+const { emails: { additionalEmails, primaryEmail } } = loadState('settings', 'personalInfoParameters', {})
```
That could prevent an exception

```suggestion
const { emails: { additionalEmails, primaryEmail } } = loadState('settings', 'personalInfoParameters', { emails: {}})
```

Pytal replied:
Wrote this similarly to other `loadState` calls in the codebase which don't handle errors — with your suggestion other errors will throw further down the chain. I've written up a potential solution which catches `loadState` errors, but would need more input from others to decide whether it's a good solution
```js
// main-personal-info.js
const EmailView = Vue.extend(EmailSection)

try {
	const { emails: { additionalEmails, primaryEmail } } = loadState('settings', 'personalInfoParameters', {})

	new EmailView({
		propsData {
			additionalEmails,
			primaryEmail,
		},
	}).$mount('#vue-emailsection')
} catch (e) {
	logger.error('Could not load personalInfoParameters state', e)
}
```

artonge replied:
Yes, you are right, it makes sense to let `loadState` throw an error instead of creating more obscure errors down the chain. I don't even think that the extra `try/catch` is necessary.

Pytal replied:
Fail fast 😂, so is it fine with you if I keep it the same and let `loadState` throw if there are errors?

artonge replied:
Yes :smiley:

artonge commented on apps/settings/src/components/PersonalInfo/shared/HeaderBar.vue:
```diff
@@ -72,8 +94,8 @@ export default {
-		addAdditionalEmail() {
-			this.$emit('addAdditionalEmail')
+		onAddAdditional() {
+			this.$emit('addAdditional')
```
I think we have a convention on event nomenclature: https://github.com/nextcloud/nextcloud-event-bus#naming-convention

Pytal replied:
Open to suggestions :)

Pytal replied:
I think that's for global events — open to suggestions though :)

Pytal replied:
I think that's for global events — open to suggestions though :)

artonge replied:
@skjnldsv any recommendations ?

Pytal replied:
@skjnldsv we can also add an eslint rule https://eslint.vuejs.org/rules/custom-event-name-casing.html#vue-custom-event-name-casing

skjnldsv replied:
Events should be kebab-case yeah.
The naming convention is for global events yes

Pytal replied:
Roger that @skjnldsv https://github.com/nextcloud/eslint-config/pull/175 :)

Pytal replied:
Fixup for kebab-case is in :)

Pytal committed:
Vuetify
- abstract shared components
- rewrite email section

Signed-off-by: Christopher Ng <chrng8@gmail.com>

Pytal commented on apps/settings/src/components/PersonalInfo/DisplayNameSection/DisplayName.vue:
```diff
@@ -0,0 +1,179 @@
+			required="true"
+			@input="onDisplayNameChange">
+
+		<div class="displayname-actions-container">
```
BEM'd it :)

Pytal committed:
Functional paradigm shift

Signed-off-by: Christopher Ng <chrng8@gmail.com>

Pytal commented:
/compile /

Pytal commented:
/compile amend /

Pytal commented:
/compile amend /

Pytal commented:
/compile /

Pytal commented:
/compile /

skjnldsv commented on apps/settings/src/components/PersonalInfo/DisplayNameSection/DisplayNameSection.vue:
```diff
@@ -0,0 +1,100 @@
+	<form
+		ref="form"
+		class="section"
+		@submit.stop.prevent="() => {}">
```
Does it even make sense to have a form if we're not using the form submit methods? :)

Pytal replied:
`form` has been removed, along with other validation improvements as we no longer use `checkValidity` from HTMLFormElement :)

skjnldsv commented on apps/settings/src/components/PersonalInfo/EmailSection/Email.vue:
```diff
@@ -253,6 +282,10 @@ export default {
 			}
 		},
 
+		isValid(email) {
+			return /^\S+$/.test(email)
+		},
```
I'm sure we have more than this to check against emails strings :wink:

Pytal replied:
Validation improved :)

Pytal replied:
Validation improved based on `FILTER_VALIDATE_EMAIL` :) https://github.com/nextcloud/server/blob/079f76ab12730eb9ea107bac9b4041cd136f2446/apps/provisioning_api/lib/Controller/UsersController.php#L841

skjnldsv commented:
Good for me, other tiny comments can be addressed in follow-up :)

nickvergessen commented on apps/settings/src/components/PersonalInfo/shared/HeaderBar.vue:
```diff
@@ -21,44 +21,66 @@
-		<label for="email">
-			{{ t('settings', 'Email') }}
+		<label :for="labelFor">
+			{{ t('settings', accountProperty) }}
```
This does not work for translations (unless it is garanteed that the account property string is still being translated in another place)

nickvergessen commented on apps/settings/src/constants/AccountPropertyConstants.js:
```diff
@@ -26,24 +26,48 @@
+
+/** Enum of account properties to human readable account properties */
+export const ACCOUNT_PROPERTY_READABLE_ENUM = Object.freeze({
+	ADDRESS: 'Address',
+	AVATAR: 'Avatar',
+	DISPLAYNAME: 'Full name',
+	EMAIL: 'Email',
+	EMAIL_COLLECTION: 'Additional Email',
+	PHONE: 'Phone',
+	TWITTER: 'Twitter',
+	WEBSITE: 'Website',
```
```suggestion
	ADDRESS: t('settings', 'Address'),
	AVATAR: t('settings', 'Avatar'),
	DISPLAYNAME: t('settings', 'Full name'),
	EMAIL: t('settings', 'Email'),
	EMAIL_COLLECTION: t('settings', 'Additional Email'),
	PHONE: t('settings', 'Phone'),
	TWITTER: t('settings', 'Twitter'),
	WEBSITE: t('settings', 'Website'),
```

Pytal replied:
Translated strings in :)

Pytal replied:
Translated strings are in :)

Pytal committed:
Refine input validation
- Remove usage of JS core checkValidity() in favour of custom backend compliant validation
- Rewrite and refactor with removal of form tag in favour of section
- Scope styles
- Remove many uses of $nextTick
- Refine disabled state logic
- Translate account property constants

Signed-off-by: Christopher Ng <chrng8@gmail.com>

Pytal commented:
/compile /

Pytal commented:
/compile /

nextcloud-command committed:
Compile assets

Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>
